### ISSUE TITLE
Import of Hipchat export (tar.gz) is failing with invalid format

### ISSUE THREAD

**Issue by icerootsvw** [RocketChat/Rocket.Chat#11355]
Rocket.Chat Version: 0.65.1  (snap installation)
Host-System: Debian GNU Linux Stretch

I created an export of my current Hipchat Installation (Atlassian HipChat 2.2.9 (2018.01.23.110010)). This is a standalone installation and not the cloud version.

The export was an AES encrypted file (around 950MB)
hipchat-2018-07-05_09-30-12.tar.gz.aes: openssl enc'd data with salted password

I decrypted the file with
openssl aes-256-cbc -d -in /home/mbasse/hipchat-2018-07-05_09-30-12.tar.gz.aes -out /home/mbasse/hipchat-2018-07-05_09-30-12.tar.gz  -pass pass:******

So i got the tar.gz file
hipchat-2018-07-05_09-30-12.tar.gz: gzip compressed data, was "hipchat-2018-07-05_09-30-12.tar", last modified: Thu Jul  5 09:30:12 2018, max compression

I was able to open the tar.gz file with my local file browser. I was also able to extract it.

I tried to import the file with these 2 options (because i dont know what the difference between Hipchat and Hipchat Enterprise is
FQDN/admin/import/prepare/hipchat
and also
FQDN/admin/import/prepare/hipchatenterprise

In both cases i got after some seconds the error message "Ungültiges Importformat" (english: invalidf import format)

I hope these information are useful. If you need further information please feel free to ask.

Greetings from Germany
Michael

**Comment by icerootsvw** [comment]
as i just figured out, there is a new hipchat version (2.4.0). I will update the system and will recheck the import

**Comment by icerootsvw** [comment]
even with Hipchat 2.4.0 and the export from that system i am getting the exact same error

**Comment by piotrkochan** [comment]
I'm getting a lot of errors:

```
I20180717-09:55:06.290(2)? Exception in callback of async function: SyntaxError: Unexpected token , in JSON at position 5
I20180717-09:55:06.290(2)?     at JSON.parse (<anonymous>:null:null)
I20180717-09:55:06.291(2)?     at stream.on.Meteor.bindEnvironment.chunk (/media/sdfsdfsdf/Rocket.Chat/.meteor/local/build/programs/server/packages/rocketchat_importer-hipchat-enterprise.js:125:31)
I20180717-09:55:06.291(2)?     at runWithEnvironment (packages/meteor.js:1238:24)
```

I've looked into source and debugged `packages/rocketchat-importer-hipchat-enterprise/server/importer.js`. It is failing on this `JSON.parse(chunk)`:
```
const promise = new Promise((resolve, reject) => {
			this.extract.on('entry', Meteor.bindEnvironment((header, stream, next) => {
				if (header.name.indexOf('.json') !== -1) {
					const info = this.path.parse(header.name);

					stream.on('data', Meteor.bindEnvironment((chunk) => {
						this.logger.debug(`Processing the file: ${ header.name }`);
						const file = JSON.parse(chunk);
```
because `chunk` is invalid JSON like:

```
 y": "private",       "room_admins": [
    452
      ],
  "topic": ""
    }
  },
```

Maybe it is problem with 'tar-stream' package.

**Comment by jakob-lundberg** [comment]
It looks like the importer takes chunks out of the tar stream and tries to JSON parse them. But to get correct JSON the entire file is needed at once. Maybe the the chunks first needs to be collected in a temp file and then parsed?

**Comment by piotrkochan** [comment]
But I assume it worked before
W dniu pon., 30.07.2018 o 10:19 Jakob Lundberg <notifications@github.com>
napisał(a):

**Comment by piotrkochan** [comment]
And loading json into file means it must be loaded into memory after so
file step is unnecessary. U think the better way is to use json stream
package.
W dniu pon., 30.07.2018 o 11:35 PiKey <ptrkchn@gmail.com> napisał(a):

**Comment by jakob-lundberg** [comment]
Well loading everything into memory is also possible. Here is how it could work
https://github.com/jakob-lundberg/Rocket.Chat/commit/dd770b2046c7afa2410d49e9c00f9a1d7d779085

**Comment by piotrkochan** [comment]
Still, this is very memory consuming way. I think that is would be better to use JSON Stream, something in the way like:

```
this.extract.on('entry', Meteor.bindEnvironment((header, stream, next) => {
[....]
stream
    .pipe(this.JSONStream.parse('*'))
    .on('data', Meteor.bindEnvironment((chunk) => {
        console.log(chunk); // {User: ...} or {Room: ...} etc
    })
```

In this approach tar-stream streams file contents while Json Strem tries to assemble them on the fly. "header" is still available so could be used to detect which file is being read.

**Comment by icerootsvw** [comment]
Is there something how i can support you (i am not a developer).
Maybe this is interesting for you. Hipchat will be canceled and Attlassian will switch to slack in 02.2019. 

https://www.theverge.com/2018/7/26/17619482/slack-hipchat-acquisition-stride-atlassian-partnership-microsoft-teams-competition


So i guess there will be more users in the next months which are migrating from Hipchat Server to Rocket.Chat (like we are doing)

**Comment by jakob-lundberg** [comment]
@piotrkochan Will you make a pull request with your solution? Otherwise I can make one with my simple "load entire file into memory" approach.

**Comment by piotrkochan** [comment]
@jakob-lundberg sorry I can't

**Comment by artem-levashov** [comment]
on my side  problem comes becuase of this  35mHipChat (tar.gz) Importer ➔ warn HipChat Enterprise importer doesn't know what to do with the file "emoticons.json" :o { root: '',   dir: '',   base: 'emoticons.json',   ext: '.json',   name: 'emoticons' } 
[34mI20180924-02:48:55.208(-5) [35mHipChat (tar.gz) Importer ➔ warn HipChat Enterprise importer doesn't know what to do with the file "metadata.json" :o { root: '',   dir: '',   base: 'metadata.json',   ext: '.json',   name: 'metadata' }

**Comment by artem-levashov** [comment]
just  found in logs that import continue to  process somehow  my file after error

**Comment by artem-levashov** [comment]
i see same error   Exception while invoking method 'prepareImport' TypeError: First argument must be a string, Buffer, ArrayBuffer, Array, or array-like object.   when i try to upload 950Mb file.    When i try to upload 24Mb then this  stage is ok. but then everything stops on following 
Sep 24 04:55:38 xloo-rchat rocketchat: rocketchat_logger rocketchat_logger.js:278 #033[34mHipChat (tar.gz) Importer ➔ debug#033[39m HipChat (tar.gz) is now at importer_preparing_started.
Sep 24 04:55:38 xloo-rchat rocketchat: rocketchat_logger rocketchat_logger.js:278 #033[34mHipChat (tar.gz) Importer ➔ debug#033[39m Processing the file: users.json
Sep 24 04:55:38 xloo-rchat rocketchat: rocketchat_logger rocketchat_logger.js:278 #033[34mHipChat (tar.gz) Importer ➔ debug#033[39m HipChat (tar.gz) is now at importer_preparing_users.
Sep 24 04:55:38 xloo-rchat rocketchat: rocketchat_logger rocketchat_logger.js:278 #033[34mHipChat (tar.gz) Importer ➔ debug#033[39m Processing the file: rooms.json
Sep 24 04:55:38 xloo-rchat rocketchat: rocketchat_logger rocketchat_logger.js:278 #033[34mHipChat (tar.gz) Importer ➔ debug#033[39m HipChat (tar.gz) is now at importer_preparing_channels.
Sep 24 04:55:38 xloo-rchat rocketchat: rocketchat_logger rocketchat_logger.js:278 #033[34mHipChat (tar.gz) Importer ➔ debug#033[39m HipChat (tar.gz) is now at importer_preparing_messages.
Sep 24 04:55:38 xloo-rchat rocketchat: rocketchat_logger rocketchat_logger.js:278 #033[35mHipChat (tar.gz) Importer ➔ warn#033[39m The loaded users count 268, the loaded rooms 217, and the loaded messages 0
Sep 24 04:55:38 xloo-rchat rocketchat: rocketchat_logger rocketchat_logger.js:278 #033[34mHipChat (tar.gz) Importer ➔ debug#033[39m HipChat (tar.gz) is now at importer_import_failed.

My small archive contains only users and room  and no massages and no attachments

**Comment by artem-levashov** [comment]
i found in the code  if (tempUsers.length === 0 || tempRooms.length === 0 || messagesCount === 0) {
... return }  

it looks like  i canot import without  messages. but with messages my file is  too large.  i will try to copy several messages from be archive to small

**Comment by piotrkochan** [comment]
I recommend upload file WITHOUT ATTACHMENTS. Rocket will download them
online

pon., 24 wrz 2018 o 14:00 artem-levashov <notifications@github.com>
napisał(a):

**Comment by artem-levashov** [comment]
with history of  only 1 room import works and i`m able now  to select what i want to import (but  unfortunatly here is also problem - Buttons  Uncheck deleted and archived  are not working - need to unchek each item manualy.)

so here is still big problem for me how to transfer  history.  my full history wihtout  attachments has  size  950Mb in archive - it is  about 3 years of using  hipchat. it looks like Rocket  do not wait while such big size will be uploaded.  I think a lot of people needs it to work - the February is close :-)

**Comment by manuelbliemel** [comment]
I'm stuck with the same problem.
Even without attachments the import does not work.
Has anybody managed to solve the issue?

**Comment by marcinkorzycki** [comment]
Bump - the same issue here.

**Comment by piotrkochan** [comment]
@marcinkorzycki RocketChat 0.73.0 introduced some Hipchat Enterprise importer improvements (https://github.com/RocketChat/Rocket.Chat/pull/12985), did You tried it?

**Comment by jandeger** [comment]
I have the same problem when trying to import from Stride. I have a tar,gz file that I can see all users, rooms etc. I can select the file and in the GUI I see all users and rooms but the buttons doesn't work. Please try to fix this ASAP since only a few weeks left before HipChat/Stride is closed and Slack is not an option for us.
I've updated to latest version 0.73.2 but still the same problem.

**Comment by marcoj** [comment]
We're having exactly the same issue. any news on this? The Stride end date is coming close and Rocket is exactly what we need...

**Comment by artem-levashov** [comment]
We was able to move from hipchat to rocket only by longway workaround. We have dowkoaded history by small periods. Each file was no more then 100mb maybe less. We spent about 1.5 weeks to upload history of 2 years.

**Comment by marcoj** [comment]
Well, we're trying to import only the chat rooms + users and that's around a 2mb file... so file size is not the issue

**Comment by jandeger** [comment]
In the new version 0.74 that is coming I saw there was something fixed for HipChat import. From the support they told us that they only support import from HipChat not Stride, I think it's the same format of the export file though. Anyway, we have give up and are moving to Slack, quite good offer the first year for the paid plan.

**Comment by artem-levashov** [comment]
Check room topics. We had problem with import when russian letters was used in room name

**Comment by artem-levashov** [comment]
Antoher problem was withusers with same email but different names

**Comment by Hudell** [comment]
All known hipchat import issues have been fixed with the release of version 0.74.3. If anybody runs into any issue with it, please open a new issue.
